### PR TITLE
Implement MPS state initialization

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

With cuquantum 24.03, we can now implement MPS state initialization (with `cutensornetStateInitializeMPS`)

- Re-implement (also fixing bugs) in `MPSSimulationState::createFromStateVec`: it was a boiler-plate and was not tested since we didn't have `cutensornetStateInitializeMPS`. Now implement it properly and add a `checkAllocationFromStateVecGeneral` test, which initialize a multi-qubit random state for validation.

- Refactor a bunch of unused code from the previous `state` branch, which is now deemed unnecessary.

### CODE REVIEW

**Review by amccaskey — APPROVED** [review]

File: runtime/nvqir/cutensornet/mps_simulation_state.cpp [hunk 1]
```diff
@@ -448,15 +430,27 @@ static Eigen::MatrixXcd reshapeStateVec(const Eigen::VectorXcd &stateVec) {
...
-                                         std::size_t dataType) {
-  Eigen::VectorXcd stateVec = Eigen::Map<Eigen::VectorXcd>(
-      reinterpret_cast<std::complex<double> *>(ptr), size);
+static Eigen::MatrixXcd reshapeStateVec(const Eigen::VectorXcd &stateVec) {
+  Eigen::MatrixXcd A = stateVec;
+  A.transposeInPlace();
+  return reshapeMatrix(A);
+}
+
+std::pair<std::unique_ptr<TensorNetState>, std::vector<MPSTensor>>
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/mps_simulation_state.cpp:439]: Might be good to have a wrapper class that holds the unique_ptr.
```c++
struct MyTensorStateThingy {
  std::unique_ptr<TensorNetState> thing;
};
```
That way you can hide that there is a unique_ptr a bit from this interface. (Not sure it is worth it.)

File: runtime/nvqir/cutensornet/mps_simulation_state.cpp [hunk 2]
```diff
@@ -465,9 +459,10 @@ MPSSimulationState::createFromSizeAndPtr(std::size_t size, void *ptr,
     MPSTensor stateTensor;
     stateTensor.deviceData = d_tensor;
     stateTensor.extents = std::vector<int64_t>{2};
-
-    return std::make_unique<MPSSimulationState>(
-        std::move(state), std::vector<MPSTensor>{stateTensor}, m_cutnHandle);
+    auto state =
+        TensorNetState::createFromMpsTensors({stateTensor}, cutnHandle);
+    return std::make_pair(std::move(state),
+                          std::vector<MPSTensor>{stateTensor});
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/mps_simulation_state.cpp:465]:
```suggestion
    return {std::move(state),
                          std::vector<MPSTensor>{stateTensor}};
```
The new school syntax.

File: runtime/nvqir/cutensornet/mps_simulation_state.cpp [hunk 3]
```diff
@@ -480,14 +475,41 @@ MPSSimulationState::createFromSizeAndPtr(std::size_t size, void *ptr,
...
+    assert(U.cols() == S.size());
+    assert(V.cols() == S.size());
+    if (S.size() <= bondDim) {
+      Eigen::MatrixXcd newU(U.rows(), bondDim);
+      newU.leftCols(U.cols()) = U;
+      Eigen::MatrixXcd newV(V.rows(), bondDim);
+      newV.leftCols(V.cols()) = V;
+      Eigen::VectorXd newS(bondDim);
+      newS.head(S.size()) = S;
+      return std::make_tuple(newU, newS, newV);
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/mps_simulation_state.cpp:490]:
```suggestion
      return {newU, newS, newV};
```

File: runtime/nvqir/cutensornet/simulator_mps_register.cpp [hunk 4]
```diff
@@ -89,7 +37,21 @@ class SimulatorMPS : public SimulatorTensorNetBase {
...
+        dynamic_cast<const MPSSimulationState *>(&in_state);
+    if (!casted)
+      throw std::invalid_argument(
+          "[SimulatorMPS simulator] Incompatible state input");
+    if (!m_state) {
+      m_state = TensorNetState::createFromMpsTensors(casted->getMpsTensors(),
+                                                     m_cutnHandle);
+    } else {
+      // Expand an existing state:
+      // Append MPS tensors
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/simulator_mps_register.cpp:51]:
```suggestion
      // Expand an existing state: Append MPS tensors
```

File: runtime/nvqir/cutensornet/tensornet_state.cpp [hunk 5]
```diff
@@ -579,8 +550,15 @@ TensorNetState::createFromStateVector(std::span<std::complex<double>> stateVec,
...
   // FIXME: use CUDA toolkit, e.g., cuBlas, to construct this projector matrix.
-  auto ket =
-      Eigen::Map<const Eigen::VectorXcd>(stateVec.data(), stateVec.size());
+  Eigen::VectorXcd ket(stateVec.size());
+  for (std::size_t i = 0; i < stateVec.size(); ++i) {
+    std::bitset<64> bs(i);
+    std::string bitStr = bs.to_string();
+    std::reverse(bitStr.begin(), bitStr.end());
+    bitStr = bitStr.substr(0, numQubits);
+    ket[std::stoull(bitStr, nullptr, 2)] = stateVec[i];
```
Comment (schweitzpgi) [inline runtime/nvqir/cutensornet/tensornet_state.cpp:559]: Was this same code up above?

**Review by schweitzpgi — APPROVED** [review]
🥇

File: runtime/nvqir/cutensornet/tensornet_state.cpp (hunk 5 above)
Comment (1tnguyen) [reply runtime/nvqir/cutensornet/tensornet_state.cpp]: Yes, it's a duplicate. Fixed in https://github.com/NVIDIA/cuda-quantum/pull/1630/commits/97532463b0e9d698433ae2b018c818fca8681bcc by creating a common helper.

**Pull request merged**